README.md
# Payture InPay API

> For node.js and browser

## 📦 Install

```sh
$ yarn add @strelka/payture-api
```

## 📖 [API](./docs/API.md)

## 🔗 [Payture Offical Docs](https://payture.com/api)

Strelka


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v1.1.1](https://github.com/strelka-institute/payture-api/compare/v1.1.0...v1.1.1) - 2018-09-12

### Fixed

- [`d862891`](https://github.com/strelka-institute/payture-api/commit/d8628918ed348ffb6fa7beef51bb3c58645e75fd) 🚑 Fix missing `Language` field

## [v1.1.0](https://github.com/strelka-institute/payture-api/compare/v1.0.0...v1.1.0) - 2018-09-12
### Changed

- [`bb8efc9`](https://github.com/strelka-institute/payture-api/commit/bb8efc9a8316c370d84385bd65110a78d4114e43) 🔥 Remove `opts.chequeContactEmail`

Use `Cheque` object in `api.init` and `api.getWidgetUrl`

### Added

- [`d9e8b61`](https://github.com/strelka-institute/payture-api/commit/d9e8b61e48b1f68dab5d5e9220dfe693c2b56421) ✨ Check hook `Notification` in `api.serverNotification`

### Docs & Tests

- [`ebe3b68`](https://github.com/strelka-institute/payture-api/commit/ebe3b68a0c9eee885c593947136f5de8f5a962a3) 📝 Update docs
- [`061cf5c`](https://github.com/strelka-institute/payture-api/commit/061cf5cc78c860d413106ed4bf7e62dcf003dc73) 📝 Add `cancel` to `widgetStatus` docs
- [`2835047`](https://github.com/strelka-institute/payture-api/commit/2835047d6f900ec45f3ca837f73daf3603ddafc2) 📝 Update `serverNotification` docs

## v1.0.0 - 2018-09-07
### Changed

- [`15d2f9b`](https://github.com/strelka-institute/payture-api/commit/15d2f9bdadc6246489ed53a01452c600fd8b6121) 📦 Update package info

### Added

- [`c9b0fc6`](https://github.com/strelka-institute/payture-api/commit/c9b0fc65aa867b56d3a32c1f4df251b6c3d86467) ✨ Watch widget status
- [`991a06d`](https://github.com/strelka-institute/payture-api/commit/991a06d79f4341a795f1f2a7299a76a7148c85a1) ✨ Get Widget url

### Improved

- [`af7705b`](https://github.com/strelka-institute/payture-api/commit/af7705b57b4d0d1bef80ea272f247b0a53fc8700) ♻️ Replace extend with spread, change axios instance

### Dependencies

- [`9d44340`](https://github.com/strelka-institute/payture-api/commit/9d443409b70954a1e38095aa3ba346a6d1eb2658) ➖ Remove flow
- [`fa62304`](https://github.com/strelka-institute/payture-api/commit/fa6230413eb7177d53fc492dcd1556c2aaa50249) ➖ Remove unused deps
- [`09afc09`](https://github.com/strelka-institute/payture-api/commit/09afc094bf24d6266f568e2ca5984e2cf5a3cdd0) ➕ Add `lint-staged`
- [`4c3c74f`](https://github.com/strelka-institute/payture-api/commit/4c3c74fd1f1b7120fb3fdca8527b20267fb34311) ➖ Remove `qs`

### Docs & Tests

- [`1f08765`](https://github.com/strelka-institute/payture-api/commit/1f08765e91c5e1adbcb0e3e3e9dff8a8c0ee3570) 📝 Add `serverNotification` docs
- [`58f7b35`](https://github.com/strelka-institute/payture-api/commit/58f7b356873b90af1fee99dd7894758f3483de42) 📝 Update docs
- [`9efe850`](https://github.com/strelka-institute/payture-api/commit/9efe8504900f09d4111fb0341245e846d54171a2) 📝 Add payture `Init` docs
- [`49e41a5`](https://github.com/strelka-institute/payture-api/commit/49e41a51f61fdb6bd6745d355c351a5beb6709a3) 📝 Add placeholder API.md


test/index.js
import test from 'ava'
import nanoid from 'nanoid'
import createPaytureApi from '../src'

const api = createPaytureApi()
const PAID_ORDER_ID = 'WqNl4LDHnv5250Ng8zaTQ'

test('API Init', async (t) => {
  const OrderId = nanoid()

  const res = await api.init({
    OrderId: OrderId,
    Amount: 50000,
    Product: 'ticket',
    Description: 'MyTestTransaction',
    Language: 'EN'
  })

  t.is(res.OrderId, OrderId)
  t.true(res.SessionId != null)
  t.true(res.PaymentUrl != null)

  const status = await api.status(OrderId)
  t.false(status.isPaid)
})

test('API Cheque Init', async (t) => {
  const OrderId = nanoid()

  const res = await api.init({
    OrderId: OrderId,
    Amount: 50000,
    Product: 'ticket',
    Description: 'MyTestTransaction',
    Language: 'EN',
    Cheque: {
      Positions: [
        {
          Quantity: 1.000,
          Price: 50000,
          Tax: 6,
          Text: 'Test Good'
        }
      ],
      CheckClose: { TaxationSystem: 6 },
      Message: 'Test Cheque Message'
    }
  })

  t.is(res.OrderId, OrderId)
  t.true(res.SessionId != null)
  t.true(res.PaymentUrl != null)

  const status = await api.status(OrderId)
  t.false(status.isPaid)

  const page = await api.pay(res.SessionId)
  t.true(typeof page.html === 'string' && page.html.length > 0)
})

test('API Status', async (t) => {
  const status = await api.status(PAID_ORDER_ID)
  t.true(status.isPaid)
})

test('Widget', async (t) => {
  const widgetUrl = api.getWidgetUrl({
    Amount: 50000,
    Product: 'ticket',
    Description: 'MyTestTransaction',
    Language: 'EN',
    Cheque: {
      Positions: [
        {
          Quantity: 1.000,
          Price: 50000,
          Tax: 6,
          Text: 'Test Good'
        }
      ],
      CheckClose: { TaxationSystem: 6 },
      Message: 'Test Cheque Message'
    }
  })

  const page = await api.axios.get(widgetUrl)

  t.true(typeof page.html === 'string' && page.html.length > 0)
})


src/utils.js
const xmlLibrary = require('xml-library')
const { IS_SERVER } = require('./constants')

const encodeBase64 = (src) => IS_SERVER
  ? Buffer.from(src).toString('base64')
  : window.btoa(src)

const decodeBase64 = (src) => IS_SERVER
  ? Buffer.from(src, 'base64').toString()
  : window.atob(src)

const json = (obj) => JSON.stringify(obj)

const parseXML = (str) => new Promise((resolve, reject) =>
  xmlLibrary.parseXML(str, (err, data) => {
    if (err) {
      return reject(err)
    }
    resolve(data)
  })
)

const serialize = (obj, separator = '&') =>
  Object
    .keys(obj)
    .reduce((acc, key) => {
      const value = obj[key]
      if (value == null) return acc

      return acc.concat(encodeURIComponent(key) + '=' + encodeURIComponent(value))
    }, [])
    .join(separator)

const createDeferredPromise = () => {
  const result = {}

  result.promise = new Promise((resolve, reject) => {
    result.resolve = resolve
    result.reject = reject
    result.cancel = () => resolve(new Promise(() => null))
  })

  return result
}

module.exports = {
  encodeBase64,
  decodeBase64,
  json,
  parseXML,
  serialize,
  createDeferredPromise
}


src/constants.js
const IS_SERVER = typeof window === 'undefined'

const SUCCESS_TRUE = 'True'
const SUCCESS_FALSE = 'False'
const ERROR_CODE_NONE = 'NONE'

const ROUTE_INPAY = '/apim'
const ROUTE_INIT = ROUTE_INPAY + '/Init'
const ROUTE_PAY = ROUTE_INPAY + '/Pay'
const ROUTE_STATUS = ROUTE_INPAY + '/PayStatus'

const SANDBOX_HOST = 'https://sandbox.payture.com'
const WIDGET_HOST = 'https://merchantgateway.payture.com'

const WIDGET_SUCCESS = 'CLOSE_PAYTURE_WIDGET_SUCCESS'
const WIDGET_ERROR = 'CLOSE_PAYTURE_WIDGET_ERROR'

const SUCCESS_NOTIFICATIONS = [
  'MerchantPay',
  'EngineBlockSuccess',
  'EngineChargeSuccess',
  'EnginePaySuccess'
]

const DEFAULT_OPTIONS = {
  host: SANDBOX_HOST,
  merchant: 'Merchant',
  returnUrl: 'http://example.com?orderid={orderid}&result={success}',
  widgetHost: WIDGET_HOST,
  widgetDomain: 2
}

module.exports = {
  IS_SERVER,
  DEFAULT_OPTIONS,
  SANDBOX_HOST,
  SUCCESS_TRUE,
  SUCCESS_FALSE,
  SUCCESS_NOTIFICATIONS,
  ERROR_CODE_NONE,
  ROUTE_INIT,
  ROUTE_PAY,
  ROUTE_STATUS,
  WIDGET_SUCCESS,
  WIDGET_ERROR
}


src/payture.js
const axios = require('axios')

const {
  encodeBase64,
  json,
  parseXML,
  serialize,
  createDeferredPromise
} = require('./utils')

const {
  IS_SERVER,
  SUCCESS_NOTIFICATIONS,
  WIDGET_SUCCESS,
  WIDGET_ERROR,
  DEFAULT_OPTIONS,
  SUCCESS_TRUE,
  SUCCESS_FALSE,
  ERROR_CODE_NONE,
  ROUTE_INIT,
  ROUTE_PAY,
  ROUTE_STATUS
} = require('./constants')

const isError = (data) =>
  data.Success === SUCCESS_FALSE && data.ErrCode !== ERROR_CODE_NONE

const paytureApi = (opts) => {
  const options = { ...DEFAULT_OPTIONS, ...opts }

  const api = axios.create({
    baseURL: options.host,
    headers: {
      'content-type': 'application/x-www-form-urlencoded'
    }
  })

  api.interceptors.request.use((config) => {
    if (config.method === 'post') {
      return {
        ...config,
        data: serialize(config.data)
      }
    } else if (config.method === 'get') {
      return {
        ...config,
        data: serialize({
          Key: options.merchant,
          ...config.data
        })
      }
    }

    return config
  })

  api.interceptors.response.use((res) => {
    const contentType = res.headers['content-type']

    if (contentType.indexOf('text/xml') === 0) {
      return parseXML(res.data)
        .then((node) => node.attributes)
        .then((data) => {
          if (isError(data)) {
            return Promise.reject(new Error(data.ErrCode))
          }

          return data
        })
    } else if (contentType.indexOf('text/html') === 0) {
      if (res.data.indexOf(WIDGET_ERROR) !== -1) {
        return Promise.reject(new Error(WIDGET_ERROR))
      }

      return {
        html: res.data
      }
    }

    return res.data
  })

  return api
}

/**
 * Create Payture InPay API.
 *
 * **See [Payture InPay Docs](https://payture.com/api#inpay_)**
 *
 * @param {object} [options]
 * @param {string} [options.host=https://sandbox.payture.com] — Host of your payture account
 * @param {string} [options.merchant=Merchant] — Your merchant account
 * @param {string} [options.returnUrl] — Url to return visitors after payment complete
 * @param {string} [options.widgetHost=https://merchantgateway.payture.com] — Payture widget template host
 * @param {string} [options.widgetDomain=2] — [Widget domain](https://payture.com/api#widget-docs_widget-params_) (use `1` for production)
 *
 * @returns {api}
 *
 * @example
 * const createPaytureApi = require('@strelka/payture-api')
 *
 * const api = createPaytureApi({
 *   host: 'https://sandbox.payture.com',
 *   merchant: 'Merchant',
 *   returnUrl: 'http://example.com?orderid={orderid}&result={success}',
 *   widgetDomain: 1
 * })
 *
 * // api.init(data)
 * // api.status(OrderId)
 * // api.pay(SessionId)
 * // api.getWidgetUrl(data)
 * // api.widgetStatus()
 * // api.serverNotification(res.body)
 */

const createPaytureApi = (options) => {
  const opts = { ...DEFAULT_OPTIONS, ...options }
  const api = paytureApi(opts)

  /**
   * Start [Payture InPay](https://payture.com/api#inpay_init_) session.
   *
   * @param {object} [data]
   * @param {string|number} [data.OrderId] — Unique order id
   * @param {string|number} [data.Amount] — Total price in kopeck (`1000` is `10.00₽`)
   * @param {string} [data.Product] — Product name (visible to user)
   * @param {string} [data.Total] — Price (visible to user)
   * @param {string} [data.Description] — Order description
   * @param {string} [data.Language] — Order page language `EN` or `RU`
   * @param {object} [data.Cheque] — [Cheque](https://payture.com/api#kassy-fz54_cheque-format-with-payment_) to send (optional)
   * @param {number} [data.Cheque.Message] — Order description
   * @param {object[]} [data.Cheque.Positions] — Order Items
   * @param {number} [data.Cheque.Positions.Quantity] — Quantity
   * @param {number} [data.Cheque.Positions.Price] — Price in kopeck
   * @param {number} [data.Cheque.Positions.Tax] — [Tax system](https://payture.com/api#kassy-fz54_cheque-status_) code for item
   * @param {string} [data.Cheque.Positions.Text] — Item description
   * @param {string} [data.Cheque.CheckClose]
   * @param {number} [data.Cheque.CheckClose.TaxationSystem] — [Tax system](https://payture.com/api#kassy-fz54_cheque-status_) code for order
   *
   * @return {Promise<object>} — with `{ OrderId, Amount, SessionId, PaymentUrl }`
   * @return {Promise<Error>} — with [Payture `ErrorCode`](https://payture.com/api#error-codes_) in message
   *
   * @example
   * api.init({
   *   OrderId: OrderId,
   *   Amount: 50000,
   *   Product: 'ticket',
   *   Description: 'MyTestTransaction',
   *   Language: 'EN'
   * })
   * .then(({ SessionId, PaymentUrl }) => {
   *   // Somehow send PaymentUrl to user
   * })
   * .catch((error) => {
   *   // ErrorCode in error.message
   * })
   */

  const init = ({
    OrderId,
    Cheque,
    Amount,
    Total,
    Product,
    Language,
    ...rest
  } = {}) =>
    api.post(ROUTE_INIT, {
      Key: opts.merchant,
      Data: serialize({
        SessionType: 'Pay',
        Url: opts.returnUrl,
        OrderId,
        Amount,
        Product, // Displayed in form
        Language: Language == null ? null : Language.toUpperCase(),
        Total: Total == null ? Amount / 100 : Total, // Displayed in form
        ...rest,
        ...Cheque != null && {
          Cheque: encodeBase64(json(Cheque))
        }
      }, '; ')
    }).then((res) => ({
      OrderId: res.OrderId,
      Amount: res.Amount,
      SessionId: res.SessionId,
      PaymentUrl: opts.host + ROUTE_PAY + '?SessionId=' + res.SessionId
    }))

  /**
   * Check order status.
   *
   * @param {string} [OrderId] — Used in {@link init} order
   *
   * @return {object} `result`
   * @return {boolean} `result.isPaid` — Parsed order payment status
   * @return {object} `result.raw` — Raw result from Payture
   * @return {Promise<Error>} — with [Payture `ErrorCode`](https://payture.com/api#error-codes_) in message
   *
   * @example
   *
   * api.status('WqNl4LDHnv5250Ng8zaTQ')
   *  .then((res) => {
   *    if (res.isPaid) {
   *      // Do something
   *    }
   *  })
   *  .catch((error) => {
   *    // Payment is ended with error
   *  })
   */

  const status = (OrderId) =>
    api.get(ROUTE_STATUS, { data: { OrderId } })
      .then((res) => ({
        isPaid: res.Success === SUCCESS_TRUE,
        raw: res
      }))

  const pay = (SessionId) =>
    api.post(ROUTE_PAY, { SessionId }, {
      maxRedirects: 0,
      validateStatus: null
    })

  /**
   * Create order widget url. Url is used in `<iframe>` `src`.
   *
   * @param {object} [order] — See order description in {@link init}.
   *
   * @return {string} — Url to widget page
   *
   * @example
   *
   * const widgetSrc = api.getWidgetUrl({
   *   Amount: 50000,
   *   Product: 'ticket',
   *   Description: 'MyTestTransaction',
   *   Language: 'EN',
   *   Cheque: {
   *     Positions: [
   *       {
   *         Quantity: 1.000,
   *         Price: 50000,
   *         Tax: 6,
   *         Text: 'Test Good'
   *       }
   *     ],
   *     CheckClose: { TaxationSystem: 6 },
   *     Message: 'Test Cheque Message'
   *   }
   * })
   *
   * @example
   *
   * <iframe src={widgetSrc} frameBorder={0} onLoad={handleWidgetLoad} />
   */

  const getWidgetUrl = ({
    Domain = opts.widgetDomain,
    Key = opts.merchant,
    Amount,
    Product,
    Cheque,
    Session,
    ...rest
  } = {}) => `${opts.widgetHost}?${serialize({
    domain: Domain,
    key: Key,
    amount: Amount,
    product: Product,
    session: Session,
    customParams: rest == null ? null : json(rest),
    chequeParams: Cheque == null ? null : json(Cheque)
  })}`

  /**
   * Watch for [widget order status](https://payture.com/api#widget-docs_workflow_).
   * Works in browser. Resolves promise if payment is succesful and rejects on error.
   *
   * @return {Promise} `status`
   * @return {function} `status.cancel` — cancel promise and stop event listener
   *
   * @example
   *
   * <iframe src={widgetSrc} frameBorder={0} onLoad={handleWidgetLoad} />
   *
   * @example
   *
   * const status = widgetStatus()
   *  .then((event) => {
   *    // Success: event.data === 'CLOSE_PAYTURE_WIDGET_SUCCESS'
   *  })
   *  .catch((event) => {
   *    // Error: : event.data === 'CLOSE_PAYTURE_WIDGET_ERROR'
   *  })
   *
   * status.cancel() // Cancel promise and stop listening for widget events
   */

  const widgetStatus = () => {
    if (IS_SERVER) {
      return Promise.reject(new Error('This function must be used in browser'))
    }

    const p = createDeferredPromise()

    function handleWigetEvent (event) {
      if (event.data === WIDGET_SUCCESS) {
        p.resolve(event)
        removeEventListener()
      } else if (event.data === WIDGET_ERROR) {
        p.reject(event)
        removeEventListener()
      }
    }

    window.addEventListener('message', handleWigetEvent)

    function removeEventListener () {
      window.removeEventListener('message', handleWigetEvent)
    }

    function cancel () {
      removeEventListener()
      p.cancel()
      return p.promise
    }

    return Object.assign(p.promise, { cancel })
  }

  /**
   * Check [Payture Server Notification](https://payture.com/api#notifications_).
   * Resolves promise with data on success and rejects [Payture `ErrorCode`](https://payture.com/api#error-codes_) on error.
   *
   * @param {object} [data] Parsed server POST body
   * @return {Promise}
   *
   * @example
   * const express = require('express')
   * const bodyParser = require('body-parser')
   * const paytureApi = require('@strelka/payture-api')()
   *
   * app.use('/payture/notification', bodyParser.urlencoded(), (req, res) => {
   *   paytureApi.serverNotification(req.body)
   *    .then((data) => {
   *      // Payment successful
   *    })
   *    .catch((error) => {
   *      // Something wrong happend
   *    })
   * })
   */

  const serverNotification = (data = {}) => {
    if (isError(data)) {
      const error = new Error(data.ErrCode)
      error.data = data

      return Promise.reject(error)
    }

    if (SUCCESS_NOTIFICATIONS.indexOf(data.Notification) === -1) {
      const error = new Error(data.Notification)
      error.data = data
      return Promise.reject(error)
    }

    return Promise.resolve({
      TransactionDate: new Date(data.TransactionDate),
      ...data
    })
  }

  return {
    axios: api,
    init,
    pay,
    status,
    getWidgetUrl,
    widgetStatus,
    serverNotification
  }
}

module.exports = createPaytureApi
